File: test/harness/hookensure.test.ts
// Port of harness-wrapper's pkg/harness/hookensure_test.go for the TS
// hook-install primitives (src/harness/internal/hookensure.ts).

import { describe, expect, test } from "vitest"
import { spawn, spawnSync } from "node:child_process"
import { existsSync, mkdtempSync, mkdirSync, readFileSync, statSync, utimesSync, writeFileSync } from "node:fs"
import { tmpdir } from "node:os"
import { dirname, join } from "node:path"
import { fileURLToPath } from "node:url"

import { atomicWriteFile, isManagedHookCommand, renderHookCommand, withLockedFile } from "../../src/harness/internal/hookensure.ts"

const here = dirname(fileURLToPath(import.meta.url))
const lockWorker = join(here, "fixtures", "lock-worker.ts")

function tmpDir(prefix: string): string {
  return mkdtempSync(join(tmpdir(), prefix))
}

describe("renderHookCommand / isManagedHookCommand structure", () => {
  test("rendered command has the expected shape", () => {
    const cmd = renderHookCommand(["/abs/loom", "hooks"], "claude", "stop", "loom")
    for (const want of [
      "sh -c ",
      'test -n "$HW_EVENT_SPOOL" || exit 0',
      "exec ",
      "/abs/loom",
      "hooks",
      "claude",
      "stop",
      "# meta-harness-hook:loom",
    ]) {
      expect(cmd).toContain(want)
    }
    expect(isManagedHookCommand(cmd)).toBe(true)
  })

  test("a user hook is not recognized as managed", () => {
    expect(isManagedHookCommand("sh -c 'echo user hook'")).toBe(false)
  })
})

// Runs the rendered command through a REAL shell to prove (a) the env guard
// makes it INERT without HW_EVENT_SPOOL, (b) it execs the entry-point with the
// right args WITH the spool set, and (c) an entry-point path containing a
// space survives the quoting.
describe("renderHookCommand behavior (real sh -c)", () => {
  test.each(["loomdir", "loom dir with space"])("dirName=%s", (dirName) => {
    const base = join(tmpDir("hookensure-behavior-"), dirName)
    mkdirSync(base, { recursive: true })
    const out = join(base, "invoked.txt")
    const loom = join(base, "loom")
    // Fake entry-point: records its args so we can assert exec reached it.
    const script = `#!/bin/sh\nprintf '%s' "$*" > ${posixQ(out)}\n`
    writeFileSync(loom, script, { mode: 0o755 })

    const cmd = renderHookCommand([loom, "hooks"], "claude", "stop", "loom")

    // (a) No spool → guard exits 0, entry-point NOT invoked.
    runSh(cmd, {})
    expect(existsSync(out)).toBe(false)

    // (b)+(c) With spool → entry-point invoked with "hooks claude stop".
    runSh(cmd, { HW_EVENT_SPOOL: "/tmp/whatever" })
    const got = readFileSync(out, "utf8")
    expect(got).toBe("hooks claude stop")
  })
})

function runSh(command: string, extraEnv: Record<string, string>): void {
  // Clean env (just PATH) so a real HW_EVENT_SPOOL in the test runner can't
  // leak in and defeat the no-spool case.
  const result = spawnSync("sh", ["-c", command], {
    env: { PATH: "/usr/bin:/bin", ...extraEnv },
  })
  if (result.status !== 0) {
    throw new Error(`sh -c failed (status ${result.status}): ${result.stderr?.toString()}`)
  }
}

// Mirrors the module's single-quote escaping for use inside the fake
// entry-point script (which itself is run by sh).
function posixQ(s: string): string {
  return `'${s.split("'").join("'\\''")}'`
}

describe("withLockedFile", () => {
  test("no-change (fn returns null) writes nothing and leaves no target file", () => {
    const target = join(tmpDir("hookensure-nochange-"), "untouched")
    withLockedFile(target, () => null)
    expect(existsSync(target)).toBe(false)
  })

  test("the .lock sidecar is left on disk permanently after use (by design, matching Go)", () => {
    const target = join(tmpDir("hookensure-lockleftover-"), "target")
    withLockedFile(target, () => Buffer.from("hello"))
    const lockPath = `${target}.lock`
    expect(existsSync(lockPath)).toBe(true)
    // Even a no-op call creates + retains the sidecar (the lock is acquired
    // before fn ever runs).
    const target2 = join(tmpDir("hookensure-lockleftover-noop-"), "target2")
    withLockedFile(target2, () => null)
    expect(existsSync(`${target2}.lock`)).toBe(true)
  })

  test("a stale sidecar (old mtime, no live holder) is taken over rather than blocking forever", () => {
    const target = join(tmpDir("hookensure-stale-"), "target")
    mkdirSync(dirname(target), { recursive: true })
    const lockPath = `${target}.lock`
    writeFileSync(lockPath, "12345\n")
    // Backdate the sidecar well past the ~10s staleness window.
    const old = new Date(Date.now() - 60_000)
    utimesSync(lockPath, old, old)

    const start = Date.now()
    withLockedFile(target, () => Buffer.from("fresh"))
    const elapsed = Date.now() - start

    expect(readFileSync(target, "utf8")).toBe("fresh")
    // Takeover should be near-immediate, not "wait out the whole staleness window".
    expect(elapsed).toBeLessThan(5_000)
  })

  test("multi-process contention: N OS processes racing the same lock file serialize their writes", async () => {
    const target = join(tmpDir("hookensure-multiproc-"), "counter")
    const workers = 20
    const incrementsPerWorker = 1

    await Promise.all(
      Array.from({ length: workers }, () => runWorker(lockWorker, [target, String(incrementsPerWorker)])),
    )

    const data = readFileSync(target, "utf8")
    expect(Number(data.trim())).toBe(workers * incrementsPerWorker)
  }, 30_000)
})

function runWorker(script: string, args: string[]): Promise<void> {
  return new Promise((resolve, reject) => {
    const proc = spawn(process.execPath, [script, ...args], { stdio: ["ignore", "pipe", "pipe"] })
    let stderr = ""
    proc.stderr.on("data", (d) => { stderr += d })
    proc.on("error", reject)
    proc.on("close", (code) => {
      if (code !== 0) reject(new Error(`worker exited ${code}: ${stderr}`))
      else resolve()
    })
  })
}

describe("atomicWriteFile", () => {
  test("writes content and leaves no temp file behind", () => {
    const dir = tmpDir("hookensure-atomic-")
    const target = join(dir, "file.txt")
    atomicWriteFile(target, Buffer.from("content"))
    expect(readFileSync(target, "utf8")).toBe("content")
    expect(statSync(target).isFile()).toBe(true)
  })
})

File: pkg/harness/hookensure_test.go
package harness_test

import (
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"testing"

	"github.com/olesho/harness-wrapper/pkg/harness"
)

func TestRenderHookCommandStructure(t *testing.T) {
	cmd := harness.RenderHookCommand([]string{"/abs/loom", "hooks"}, "claude", "stop", "loom")
	for _, want := range []string{
		"sh -c ",
		`test -n "$HW_EVENT_SPOOL" || exit 0`,
		"exec ",
		"/abs/loom",
		"hooks",
		"claude",
		"stop",
		"# harness-wrapper-hook:loom",
	} {
		if !strings.Contains(cmd, want) {
			t.Errorf("rendered command missing %q:\n%s", want, cmd)
		}
	}
	if !harness.IsManagedHookCommand(cmd) {
		t.Error("IsManagedHookCommand should recognize a freshly rendered command")
	}
	if harness.IsManagedHookCommand(`sh -c 'echo user hook'`) {
		t.Error("a user hook must NOT be recognized as loom-managed")
	}
}

// TestRenderHookCommandBehavior runs the rendered command through a shell to
// prove (a) the env guard makes it INERT without HW_EVENT_SPOOL, (b) it execs
// loom with the right args WITH the spool set, and (c) a loom path containing a
// space survives the quoting.
func TestRenderHookCommandBehavior(t *testing.T) {
	for _, dirName := range []string{"loomdir", "loom dir with space"} {
		t.Run(dirName, func(t *testing.T) {
			checkRenderedHookBehavior(t, dirName)
		})
	}
}

// checkRenderedHookBehavior runs a rendered hook command through a shell and
// asserts the env guard, exec target, and quoting all behave.
func checkRenderedHookBehavior(t *testing.T, dirName string) {
	base := filepath.Join(t.TempDir(), dirName)
	if err := os.MkdirAll(base, 0o755); err != nil {
		t.Fatal(err)
	}
	out := filepath.Join(base, "invoked.txt")
	loom := filepath.Join(base, "loom")
	// Fake loom: record its args so we can assert exec reached it correctly.
	script := "#!/bin/sh\nprintf '%s' \"$*\" > " + posixQ(out) + "\n"
	if err := os.WriteFile(loom, []byte(script), 0o755); err != nil { //nolint:gosec // test fixture executable
		t.Fatal(err)
	}
	cmd := harness.RenderHookCommand([]string{loom, "hooks"}, "claude", "stop", "loom")

	// (a) No spool → guard exits 0, loom NOT invoked.
	runSh(t, cmd, nil)
	if _, err := os.Stat(out); !os.IsNotExist(err) {
		t.Fatal("guard failed: loom was invoked without HW_EVENT_SPOOL")
	}
	// (b)+(c) With spool → loom invoked with "hooks claude stop".
	runSh(t, cmd, []string{"HW_EVENT_SPOOL=/tmp/whatever"})
	got, err := os.ReadFile(out) //nolint:gosec // test temp path
	if err != nil {
		t.Fatalf("loom was not invoked with HW_EVENT_SPOOL set: %v", err)
	}
	if string(got) != "hooks claude stop" {
		t.Errorf("loom args = %q, want %q", got, "hooks claude stop")
	}
}

func runSh(t *testing.T, command string, extraEnv []string) {
	t.Helper()
	c := exec.Command("sh", "-c", command)
	// Start from a CLEAN env so a real HW_EVENT_SPOOL in the test runner can't
	// leak in and defeat the no-spool case.
	c.Env = append([]string{"PATH=/usr/bin:/bin"}, extraEnv...)
	if out, err := c.CombinedOutput(); err != nil {
		t.Fatalf("sh -c failed: %v\n%s", err, out)
	}
}

// posixQ mirrors the package's single-quote escaping for use inside the fake
// loom script (which itself is run by sh).
func posixQ(s string) string { return "'" + strings.ReplaceAll(s, "'", `'\''`) + "'" }

func TestWithLockedFileSerializesConcurrentWriters(t *testing.T) {
	// N goroutines each read-modify-write the same file under the lock; without
	// serialization, lost updates would make the final count < N.
	target := filepath.Join(t.TempDir(), "counter")
	const n = 30
	var wg sync.WaitGroup
	for i := 0; i < n; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			err := harness.WithLockedFile(target, func(existing []byte) ([]byte, error) {
				cur := 0
				if len(existing) > 0 {
					cur, _ = strconv.Atoi(strings.TrimSpace(string(existing)))
				}
				return []byte(strconv.Itoa(cur + 1)), nil
			})
			if err != nil {
				t.Errorf("WithLockedFile: %v", err)
			}
		}()
	}
	wg.Wait()

	data, err := os.ReadFile(target) //nolint:gosec // test temp path
	if err != nil {
		t.Fatal(err)
	}
	if got, _ := strconv.Atoi(strings.TrimSpace(string(data))); got != n {
		t.Fatalf("counter = %d, want %d (lost updates ⇒ lock not serializing)", got, n)
	}
}

func TestWithLockedFileNoChange(t *testing.T) {
	// Returning (nil, nil) writes nothing and leaves no file behind.
	target := filepath.Join(t.TempDir(), "untouched")
	if err := harness.WithLockedFile(target, func([]byte) ([]byte, error) { return nil, nil }); err != nil {
		t.Fatal(err)
	}
	if _, err := os.Stat(target); !os.IsNotExist(err) {
		t.Error("no-change should not create the target file")
	}
}
